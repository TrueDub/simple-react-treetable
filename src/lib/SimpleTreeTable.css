.shown {

}

.hidden {
    display: none;
}

.iconPadding {
    padding-left: 0.5em;
}

.buttonAsLink {
    background: none !important;
    color: inherit;
    border: none;
    padding: 0 !important;
    font: inherit;
    /*border is optional*/
    border-bottom: 1px solid #444;
    cursor: pointer;
}

.topbar-container {
    display: grid;
    grid-template-columns: 25% 25% 25% 25%;
}

.count-area {
    text-align: right;
    vertical-align: middle;
}