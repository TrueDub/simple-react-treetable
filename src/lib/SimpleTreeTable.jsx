import React from 'react';
import PropTypes from 'prop-types';
import {FontAwesomeIcon} from '@fortawesome/react-fontawesome';
import {faAngleRight, faAngleDown, faAngleUp} from '@fortawesome/free-solid-svg-icons';

import './SimpleTreeTable.css';

class SimpleTreeTable extends React.Component {

    constructor(props) {
        super(props);
        let initialState = this.generateInitialState();
        this.state = {
            enhancedTableData: initialState.enhancedTableData,
            expanded: false,
            enhancedColumns: initialState.enhancedColumns,
            showResetSortingButton: initialState.showResetSortingButton
        };
    }

    generateInitialState() {
        let visibleRows = this.props.control.hasOwnProperty('visibleRows') ? this.props.control.visibleRows : 1;
        let enhancedTableData = this.generateStateTableData(this.props.tableData, visibleRows);
        let enhancedColumns = this.generateColumnState(this.props.columns);
        let initialSortField = null;
        let initialSortOrder = null;
        let showResetSortingButton = false;
        enhancedColumns.forEach(column => {
            if (column.sortOrder !== 'none') {
                initialSortField = column.dataField;
                initialSortOrder = column.sortOrder;
                showResetSortingButton = true;
            }
        });
        if (initialSortField !== null) {
            enhancedTableData = this.sortBy(enhancedTableData, initialSortField, initialSortOrder);
        }
        return {
            enhancedTableData: enhancedTableData,
            enhancedColumns: enhancedColumns,
            showResetSortingButton: showResetSortingButton
        };
    }

    generateStateTableData(tree, visibleRows) {
        let n = 1;
        return (function recurse(children, parent = 0, rowLevel = 1) {
            if (children) {
                return children.map(node => {
                    let rowID = n++;
                    return Object.assign({}, node, {
                        rowID: rowID,
                        rowLevel: rowLevel,
                        parentRowID: parent,
                        visible: rowLevel <= visibleRows,
                        expanded: rowLevel < visibleRows,
                        children: recurse(node.children, rowID, rowLevel + 1)
                    })
                });
            }
        })(tree);
    }

    generateColumnState(initialColumns) {
        return initialColumns.map(node => {
            let sortOrder = node.hasOwnProperty('sortOrder') ? node.sorted : 'none';
            return Object.assign({}, node, {
                sortable: node.hasOwnProperty('sortable') ? node.sortable : true,
                sortOrder: sortOrder
            })
        });
    }

    expandOrCollapseAll() {
        let action = !this.state.expanded;
        let newTree = (function recurse(children) {
            return children.map(node => {
                let visibleAction = node.rowLevel === 1 ? true : action;
                return Object.assign({}, node, {
                    visible: visibleAction,
                    expanded: action,
                    children: recurse(node.children)
                })
            });
        })(this.state.enhancedTableData);
        this.setState({
            enhancedTableData: newTree,
            expanded: action
        });
    }

    rowExpandOrCollapse(selectedRowID) {
        let newTree = this.expandOrCollapseTree(this.state.enhancedTableData, selectedRowID, false, false);
        this.setState({enhancedTableData: newTree});
    }

    expandOrCollapseTree(data, selectedRowID, expandAll, collapseAll) {
        return (function recurse(children, expandBranch = expandAll, collapseBranch = collapseAll) {
            return children.map(node => {
                let setExpanded = node.rowID === selectedRowID ? !node.expanded : node.expanded;
                let setVisible = node.parentRowID === selectedRowID ? !node.visible : node.visible;
                if (expandBranch) {
                    setExpanded = true;
                    setVisible = true;
                }
                if (collapseBranch) {
                    setExpanded = false;
                    setVisible = false;
                }
                //collapse and hide all below
                if (node.parentRowID === selectedRowID && !setVisible) {
                    collapseBranch = true;
                }
                return Object.assign({}, node, {
                    visible: setVisible,
                    expanded: setExpanded,
                    children: recurse(node.children, expandBranch, collapseBranch)
                })
            });
        })(data);
    }

    sortByField(fieldName) {
        let sortStatus = this.getSortStatus(fieldName);
        let sortOrder = 'asc';
        if (sortStatus === 'asc') {
            sortOrder = 'desc';
        }
        let newTree = this.sortBy(this.state.enhancedTableData, fieldName, sortOrder);
        let newColumns = this.state.enhancedColumns.map(node => {
            let newSortOrder = 'none';
            if (node.dataField === fieldName) {
                newSortOrder = sortOrder;
            }
            return Object.assign({}, node, {
                sortOrder: newSortOrder
            })
        });
        this.setState({
            enhancedTableData: newTree,
            enhancedColumns: newColumns,
            showResetSortingButton: true
        });
    }

    sortBy(data, fieldName, direction) {
        data.forEach(entry => {
            if (entry.children && entry.children.length > 0) {
                entry.children = this.sortBy(entry.children, fieldName, direction);
            }
        });
        if (direction === 'asc') {
            return data.sort((a, b) => {
                return a.data[fieldName] < b.data[fieldName] ? -1 : a.data[fieldName] > b.data[fieldName] ? 1 : 0;
            });
        } else {
            return data.sort((b, a) => {
                return a.data[fieldName] < b.data[fieldName] ? -1 : a.data[fieldName] > b.data[fieldName] ? 1 : 0;
            });
        }
    };

    getSortStatus(fieldName) {
        for (let i = 0; i < this.state.enhancedColumns.length; i++) {
            if (this.state.enhancedColumns[i].dataField === fieldName) {
                return this.state.enhancedColumns[i].sortOrder;
            }
        }
    }

    resetSorting() {
        let initialState = this.generateInitialState();
        this.setState({
            enhancedTableData: initialState.enhancedTableData,
            enhancedColumns: initialState.enhancedColumns,
            showResetSortingButton: initialState.showResetSortingButton
        });
    }

//from here down the functions deal with rendering

    generateTableBody(tableData) {
        let tableBody = [];
        tableData.forEach((dataRow) => {
                let rowData = this.processDataRow(dataRow);
                let key = dataRow.parentRowID + '-' + dataRow.rowID;
                let rowClass = dataRow.visible ? 'shown' : 'hidden';
                tableBody.push(<tr className={rowClass} key={key}>{rowData}</tr>);
                if (dataRow.children) {
                    tableBody.push(...this.generateTableBody(dataRow.children));
                }
            }
        );
        return tableBody;
    }

    generateExpandColumn(dataRow, key, dataField) {
        if (dataRow.children && dataRow.children.length > 0) {
            let iconCell = <FontAwesomeIcon icon={faAngleRight} fixedWidth
                                            onClick={this.rowExpandOrCollapse.bind(this, dataRow.rowID)}/>;
            if (dataRow.expanded) {
                iconCell = <FontAwesomeIcon icon={faAngleDown} fixedWidth
                                            onClick={this.rowExpandOrCollapse.bind(this, dataRow.rowID)}/>;
            }
            if (this.state.enhancedColumns[0].fixedWidth) {
                return (<td key={key} className={this.state.enhancedColumns[0].styleClass}
                            width={this.state.enhancedColumns[0].percentageWidth + '%'}><span
                        style={{marginLeft: dataRow.rowLevel + 'em'}}>{iconCell}<span
                        className="iconPadding">{dataRow.data[dataField]}</span></span></td>
                );
            } else {
                return (<td key={key} className={this.state.enhancedColumns[0].styleClass}><span
                        style={{marginLeft: dataRow.rowLevel + 'em'}}>{iconCell}<span
                        className="iconPadding">{dataRow.data[dataField]}</span></span></td>
                );
            }
        } else {
            if (this.state.enhancedColumns[0].fixedWidth) {
                return (
                    <td key={key} className={this.state.enhancedColumns[0].styleClass}
                        width={this.state.enhancedColumns[0].percentageWidth + '%'}><span
                        style={{marginLeft: (dataRow.rowLevel + 1.25) + 'em'}}>
                    <span className="iconPadding">{dataRow.data[dataField]}</span>
                </span>
                    </td>);
            } else {
                return (
                    <td key={key} className={this.state.enhancedColumns[0].styleClass}><span
                        style={{marginLeft: (dataRow.rowLevel + 1.25) + 'em'}}>
                    <span className="iconPadding">{dataRow.data[dataField]}</span>
                </span>
                    </td>);
            }
        }
    }

    processDataRow(dataRow) {
        let rowBody = [];
        rowBody.push(this.state.enhancedColumns.map((column, index) => {
                let key = dataRow.parentRowID + '-' + dataRow.rowID + '-' + index;
                let output = dataRow.data[column.dataField];
                if (column.renderer) {
                    output = this.state.enhancedColumns[index].renderer(dataRow, column.dataField);
                }
                if (index === 0) {
                    return this.generateExpandColumn(dataRow, key, column.dataField);
                } else {
                    if (column.fixedWidth) {
                        return (<td key={key} className={column.styleClass}
                                    width={column.percentageWidth + '%'}>{output}</td>)
                    } else {
                        return (
                            <td key={key} className={column.styleClass}>{output}</td>)
                    }
                }
            }
        ));
        return rowBody;
    }

    generateHeaderRow() {
        let headingRows = [];
        if (this.state.enhancedColumns) {
            headingRows.push(this.state.enhancedColumns.map((column) => {
                    let fieldTitle = column.heading ? column.heading : column.dataField;
                    let sortIcon = null;
                    if (column.sortOrder === 'asc') {
                        sortIcon = <FontAwesomeIcon icon={faAngleUp} fixedWidth pull="right"/>;
                    } else if (column.sortOrder === 'desc') {
                        sortIcon = <FontAwesomeIcon icon={faAngleDown} fixedWidth pull="right"/>;
                    } else {
                        sortIcon = null;
                    }
                    if (column.sortable) {
                        return <th key={fieldTitle}
                                   onClick={this.sortByField.bind(this, column.dataField)}>{sortIcon}{fieldTitle}</th>;
                    } else {
                        return <th key={fieldTitle}>{fieldTitle}</th>;
                    }
                }
            ))
        }
        return headingRows;
    }

    render() {
        let headingRows = this.generateHeaderRow();
        let tableBody = this.generateTableBody(this.state.enhancedTableData);
        return (
            <div>
                <button onClick={this.expandOrCollapseAll.bind(this)}
                        className={this.props.control.showExpandCollapseButton ? this.props.control.expandCollapseButtonClasses : 'hidden'}>
                    {this.state.expanded ? 'Collapse All' : 'Expand All'}
                </button>
                <button onClick={this.resetSorting.bind(this)}
                        className={this.state.showResetSortingButton ? this.props.control.resetSortingButtonClasses : 'hidden'}>
                    Reset Sorting
                </button>
                <table className={this.props.control.tableClasses}>
                    <thead>
                    <tr>
                        {headingRows}
                    </tr>
                    </thead>
                    <tbody>
                    {tableBody}
                    </tbody>
                </table>
            </div>
        );
    }
}

SimpleTreeTable.propTypes = {
    tableData: PropTypes.arrayOf(
        PropTypes.shape({
            data: PropTypes.object,
            children: PropTypes.arrayOf(PropTypes.object)
        })).isRequired,
    control: PropTypes.shape({
        visibleRows: PropTypes.number,
        tableClasses: PropTypes.string,
        showExpandCollapseButton: PropTypes.bool,
        expandCollapseButtonClasses: PropTypes.string,
        showResetSortingButton: PropTypes.bool,
        resetSortingButtonClasses: PropTypes.string
    }),
    columns: PropTypes.arrayOf(PropTypes.shape({
        dataField: PropTypes.string.isRequired,
        heading: PropTypes.string,
        fixedWidth: PropTypes.bool,
        percentageWidth: PropTypes.number,
        styleClass: PropTypes.string,
        renderer: PropTypes.func,
        sortable: PropTypes.bool,
        sortOrder: PropTypes.string
    }))
};

SimpleTreeTable.defaultProps = {
    tableData: [],
    control: {
        visibleRows: 1,
        tableClasses: '',
        showExpandCollapseButton: false,
        expandCollapseButtonClasses: '',
        showResetSortingButton: false,
        resetSortingButtonClasses: ''
    },
    columns: [{
        dataField: '',
        heading: '',
        fixedWidth: false,
        percentageWidth: 0,
        styleClass: '',
        renderer: null,
        sortable: true
    }]
};

export default SimpleTreeTable;