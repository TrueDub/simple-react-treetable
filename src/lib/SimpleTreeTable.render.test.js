import React from 'react';
import renderer from 'react-test-renderer';
import SimpleTreeTable from "./SimpleTreeTable";

let dataFields = ["name", "dataType", "example", "description"];
let tableData = [
    {
        data: {
            name: "name0",
            dataType: "string0",
            example: "ex0",
            description: "desc0"
        },
        children: [
            {
                data: {
                    name: "name0-0",
                    dataType: "string0-0",
                    example: "ex0-0",
                    description: "desc0-0"
                },
                children: []
            }, {
                data: {
                    name: "name0-1",
                    dataType: "string0-1",
                    example: "ex0-1",
                    description: "desc0-1"
                },
                children: []
            }, {
                data: {
                    name: "name0-2",
                    dataType: "string0-2",
                    example: "ex0-2",
                    description: "desc0-2"
                },
                children: [
                    {
                        data: {
                            name: "name0-2-1",
                            dataType: "string0-2-1",
                            example: "ex0-2-1",
                            description: "desc0-2-1"
                        },
                        children: []
                    }
                ]
            }
        ]
    },
    {
        data: {
            name: "name1",
            dataType: "string1",
            example: "ex1",
            description: "desc1"
        },
        children: []
    },
    {
        data: {
            name: "name2",
            dataType: "string2",
            example: "ex2",
            description: "desc2"
        },
        children: []
    }
];
let dataTableData = [
    {
        data: {
            name: "name0",
            dataType: "string0",
            example: "ex0",
            description: "desc0"
        }
    },
    {
        data: {
            name: "name1",
            dataType: "string1",
            example: "ex1",
            description: "desc1"
        }
    },
    {
        data: {
            name: "name2",
            dataType: "string2",
            example: "ex2",
            description: "desc2"
        }
    }
];
let controlWithButton = {
    tableClasses: "table table-bordered",
    buttonClasses: "btn btn-default pull-right",
    showButton: true
};
let columns = [
    {heading: "fred1", fixedWidth: true, percentageWidth: 25},
    {heading: "fred2", fixedWidth: true, percentageWidth: 10},
    {heading: "fred3", fixedWidth: true, percentageWidth: 25},
    {heading: "fred4", fixedWidth: true, percentageWidth: 40}
];

test('component renders as expected', () => {
    const component = renderer.create(
        <SimpleTreeTable columns={columns} dataFields={dataFields} tableData={tableData} control={controlWithButton}/>,
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});