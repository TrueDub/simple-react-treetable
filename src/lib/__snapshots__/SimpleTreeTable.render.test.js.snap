// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component renders as expected 1`] = `
<div>
  <button
    className="btn btn-default pull-right"
    onClick={[Function]}
  >
    Expand All
  </button>
  <table
    className="table table-bordered"
  >
    <thead>
      <tr>
        <th>
          fred1
        </th>
        <th>
          fred2
        </th>
        <th>
          fred3
        </th>
        <th>
          fred4
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        className="shown"
      >
        <td
          width="25%"
        >
          <span
            style={
              Object {
                "marginLeft": "1em",
              }
            }
          >
            <svg
              aria-hidden="true"
              className="svg-inline--fa fa-angle-right fa-w-8 fa-fw "
              data-icon="angle-right"
              data-prefix="fas"
              onClick={[Function]}
              role="img"
              style={Object {}}
              viewBox="0 0 256 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
                fill="currentColor"
                style={Object {}}
              />
            </svg>
            <span
              className="iconPadding"
            >
              name0
            </span>
          </span>
        </td>
        <td
          width="10%"
        >
          string0
        </td>
        <td
          width="25%"
        >
          ex0
        </td>
        <td
          width="40%"
        >
          desc0
        </td>
      </tr>
      <tr
        className="hidden"
      >
        <td
          width="25%"
        >
          <span
            style={
              Object {
                "marginLeft": "3.25em",
              }
            }
          >
            <span
              className="iconPadding"
            >
              name0-0
            </span>
          </span>
        </td>
        <td
          width="10%"
        >
          string0-0
        </td>
        <td
          width="25%"
        >
          ex0-0
        </td>
        <td
          width="40%"
        >
          desc0-0
        </td>
      </tr>
      <tr
        className="hidden"
      >
        <td
          width="25%"
        >
          <span
            style={
              Object {
                "marginLeft": "3.25em",
              }
            }
          >
            <span
              className="iconPadding"
            >
              name0-1
            </span>
          </span>
        </td>
        <td
          width="10%"
        >
          string0-1
        </td>
        <td
          width="25%"
        >
          ex0-1
        </td>
        <td
          width="40%"
        >
          desc0-1
        </td>
      </tr>
      <tr
        className="hidden"
      >
        <td
          width="25%"
        >
          <span
            style={
              Object {
                "marginLeft": "2em",
              }
            }
          >
            <svg
              aria-hidden="true"
              className="svg-inline--fa fa-angle-right fa-w-8 fa-fw "
              data-icon="angle-right"
              data-prefix="fas"
              onClick={[Function]}
              role="img"
              style={Object {}}
              viewBox="0 0 256 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
                fill="currentColor"
                style={Object {}}
              />
            </svg>
            <span
              className="iconPadding"
            >
              name0-2
            </span>
          </span>
        </td>
        <td
          width="10%"
        >
          string0-2
        </td>
        <td
          width="25%"
        >
          ex0-2
        </td>
        <td
          width="40%"
        >
          desc0-2
        </td>
      </tr>
      <tr
        className="hidden"
      >
        <td
          width="25%"
        >
          <span
            style={
              Object {
                "marginLeft": "4.25em",
              }
            }
          >
            <span
              className="iconPadding"
            >
              name0-2-1
            </span>
          </span>
        </td>
        <td
          width="10%"
        >
          string0-2-1
        </td>
        <td
          width="25%"
        >
          ex0-2-1
        </td>
        <td
          width="40%"
        >
          desc0-2-1
        </td>
      </tr>
      <tr
        className="shown"
      >
        <td
          width="25%"
        >
          <span
            style={
              Object {
                "marginLeft": "2.25em",
              }
            }
          >
            <span
              className="iconPadding"
            >
              name1
            </span>
          </span>
        </td>
        <td
          width="10%"
        >
          string1
        </td>
        <td
          width="25%"
        >
          ex1
        </td>
        <td
          width="40%"
        >
          desc1
        </td>
      </tr>
      <tr
        className="shown"
      >
        <td
          width="25%"
        >
          <span
            style={
              Object {
                "marginLeft": "2.25em",
              }
            }
          >
            <span
              className="iconPadding"
            >
              name2
            </span>
          </span>
        </td>
        <td
          width="10%"
        >
          string2
        </td>
        <td
          width="25%"
        >
          ex2
        </td>
        <td
          width="40%"
        >
          desc2
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
